src/windows/file_manager_windows_impl.cc

```diff
@@ -0,0 +1,81 @@
+/*******************************************************************************
+The MIT License (MIT)
+
+Copyright (c) 2015 Alexander Zolotarev <me@alex.bio> from Minsk, Belarus
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+// Windows implementations for FileManager
+#include <windows.h>
+#include <direct.h>
+#include <stdio.h>
+
+namespace alohalytics {
+
+const char FileManager::kDirectorySeparator = '\\';
+
+namespace {
+template <class TCloseFunc>
+struct ScopedCloseFindFileHandle {
+  HANDLE handle_;
+  TCloseFunc closer_;
+  ScopedCloseFindFileHandle(HANDLE handle, TCloseFunc closer) : handle_(handle), closer_(closer) {}
+  ~ScopedCloseFindFileHandle() { closer_(handle_); }
+};
+}  // namespace
+
+void FileManager::ForEachFileInDir(std::string directory, std::function<bool(const std::string & full_path)> lambda) {
+  // Silently ignore invalid directories.
+  if (directory.empty()) {
+    return;
+  }
+  AppendDirectorySlash(directory);
+  // TODO(AlexZ): Do we need to use *W functions (and wstrings) for files
+  // processing?
+  WIN32_FIND_DATAA find_data;
+  HANDLE handle = ::FindFirstFileA((directory + "*.*").c_str(), &find_data);
+  if (handle == INVALID_HANDLE_VALUE) {
+    return;
+  }
+  const ScopedCloseFindFileHandle closer(handle, &::FindClose);
+  do {
+    if (!(find_data.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY)) {
+      if (!lambda(directory + find_data.cFileName)) {
+        // Stop processing if lambda has returned false.
+        return;
+      }
+    }
+  } while (::FindNextFileA(handle, &find_data) != 0);
+}
+
+int64_t FileManager::GetFileSize(const std::string & full_path_to_file) {
```

Comment:
Thanks for recommendation, in this version I’m trying to avoid most exceptions. Probably review this approach later.

> On Jun 10, 2015, at 20:10, Dima <notifications@github.com> wrote:
> 
> In src/windows/file_manager_windows_impl.cc <https://github.com/KnowSheet/Alohalytics/pull/36#discussion_r32141190>:
> 
> > +  HANDLE handle = ::FindFirstFileA((directory + "*.*").c_str(), &find_data);
> > +  if (handle == INVALID_HANDLE_VALUE) {
> > +    return;
> > +  }
> > +  const ScopedCloseFindFileHandle closer(handle, &::FindClose);
> > +  do {
> > +    if (!(find_data.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY)) {
> > +      if (!lambda(directory + find_data.cFileName)) {
> > +        // Stop processing if lambda has returned false.
> > +        return;
> > +      }
> > +    }
> > +  } while (::FindNextFileA(handle, &find_data) != 0);
> > +}
> > +
> > +int64_t FileManager::GetFileSize(const std::string & full_path_to_file) {
> I'd still recommend uint64_t. In the code above you just case it into an unsigned type, so no real gain. Make it unsigned -- and throw an exception if you really want safety here.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/KnowSheet/Alohalytics/pull/36/files#r32141190>.
>